### ISSUE TITLE
AVX512 performance

### ISSUE THREAD

**Issue by halfflat** [arbor-sim/arbor#637]
Before the gather/scatter vectorization optimizations, AVX512 was considerably faster than AVX2 on KNL. Now, AVX2 is faster. So it seems likely that our use of vector memory operations on KNL is suboptimal, and this may also explain the worse performance we see on Skylake-X when AVX512 is used instead of AVX2.

Investigate and fix if possible!

*Update:* might not actually be slower on KNL, but the speed up is much smaller than we'd expect.

**Comment by halfflat** [comment]
How much work would it be to resurrect the dynamic dispatch on constraint solution, with the single pass through memory? It would be good to compare.

To be honest, I have another reason for re-considering that approach: it greatly simplifies adding different shared storage data and layouts to the multi-core back end in both modcc and the shared state classes.

**Comment by noraabiakar** [comment]
It shouldn't be too much work. I need to modify the modcc printer and test. I'll implement it and compare.

**Comment by noraabiakar** [comment]
After re-implementing the dynamic dispatch version, I found that performance is worse on significant platforms for significant tests. VTune profiling shows almost 2x retired instructions compared to the static version. This is likely due to branch misprediction. 
I also tried to alleviate some of the memory access issues in the static version, specifically the "none" index_constraints in the compound_index_add function. A simple change insures we only read/write every distinct index in a vector once from/to memory. This gives around 20% improvement for expsyn_current. The fix is in PR #673
